Major rewrite of download_all()

valgur opened the pull request:
As evidenced by the number of open issues related to it, the `download_all()` functionality has become thoroughly inadequate for the increasingly complex LTA handling logic (and bugs) on the server side. So, I saw a need for a major rewrite and refactoring.

This PR should be fairly complete, but I have not gotten around to updating the documentation and changelog yet.

Changes in this PR:

- Moved all download-related code to a separate class called `Downloader`. The class is intended to be mostly for internal use, at least currently. The existing SentinelAPI API is kept intact and backwards-compatible. I did not see a pressing need for an API break and a major version bump.
- Rewrote `download_all()`:
  * The status of each product being downloaded is tracked more clearly with a [`DownloadStatus`](https://github.com/valgur/sentinelsat/blob/64de72bc16b327045757e1362dab2148243991b6/sentinelsat/download.py#L25-L36) enum. That status info is returned from `Downloader.download_all()` as well.
  * All GET requests during download (including LTA triggering requests) are strictly limited to `n_concurrent_dl` concurrent requests with the help of a `BoundedSemaphore`. This fixes #522.
  * LTA triggering now also waits for the products to come online and are then downloaded.
  * Triggering is now much more robust against server-side bugs (such as the frequent HTTP 500: NullPonterException, Postgres bugs, etc.). Fixes #488.
  * The number of concurrent LTA triggerings is limited to 10 by default and can be configured with `n_concurrent_trigger`.
  * Default `n_concurrent_dl` increased from 2 to 4.
  * Default LTA retry period was reduced from 600 to 60 seconds.
  * tqdm progressbars for overall download and LTA retrieval progress.
  * Added `fail_fast` option to the API and CLI. Defaults to False.
  * Cleaner logged info and progressbar display thanks to #531.
- Download progress bars now include the name of the product being downloaded.
- Downloads can be cleanly interrupted. Fixes #481.
- `SentinelProductAPI` has been merged into `SentinelAPI` and `Downloader`. An alias to `SentinelAPI` has been kept for backwards-compatibility.
- Probably fixes #528.

Overall, sentinelsat should now be actually usable for downloading archived data and the CLI download experience should be much more pleasant.
An example:
```
$ sentinelsat --sentinel 2 --start 20190309 --end 20190315 --limit 20 --download
Found 130,033 products
Will download 20 products using 4 workers
Fetching archival status: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:07<00:00,  2.71product/s]
S2A_MSIL2A_20190312T200221_N0213_R128_T14XMN_20191022T090931 (e02247f8-b74a-4f83-aaff-5aa03390b219) is in the LTA and retrieval will be triggered.
S2A_MSIL2A_20190312T200221_N0213_R128_T14XMK_20191022T090931 (907889bb-d36f-41dc-a8aa-019006aa73c5) is in the LTA and retrieval will be triggered.
S2A_MSIL2A_20190312T200221_N0213_R128_T13XED_20191022T090931 (a3313bf8-5070-40f7-aab3-16dc1e347d0d) is in the LTA and retrieval will be triggered.
S2A_MSIL2A_20190312T200221_N0213_R128_T14XNK_20191022T090931 (d54dee6f-d1d6-40ea-909b-2c5af9354d5a) is in the LTA and retrieval will be triggered.
S2A_MSIL2A_20190312T200221_N0213_R128_T13XDE_20191022T090931 (96250b2d-d46e-4444-bfb7-516bd9068f35) is in the LTA and retrieval will be triggered.
S2A_MSIL2A_20190312T200221_N0213_R128_T15XVF_20191022T090931 (ed7871d4-8382-49e8-af13-3ebcd20fc67a) is in the LTA and retrieval will be triggered.
S2A_MSIL2A_20190312T200221_N0213_R128_T12XWL_20191022T090931 (2050ea0c-0f62-434c-adc4-af3e0b997d94) is in the LTA and retrieval will be triggered.
S2A_MSIL2A_20190312T200221_N0213_R128_T13XEH_20191022T090931 (692d45de-4d1e-41df-b46c-2b41d8dc1d65) is in the LTA and retrieval will be triggered.
S2A_MSIL2A_20190312T200221_N0213_R128_T14XNM_20191022T090931 (99e93043-a721-4560-ba97-6b23c4e1907a) is in the LTA and retrieval will be triggered.
907889bb-d36f-41dc-a8aa-019006aa73c5 accepted for retrieval                                                                                                                                                                    
e02247f8-b74a-4f83-aaff-5aa03390b219 accepted for retrieval                                                                                                                                                                    
ed7871d4-8382-49e8-af13-3ebcd20fc67a accepted for retrieval                                                                                                                                                                    
96250b2d-d46e-4444-bfb7-516bd9068f35 accepted for retrieval                                                                                                                                                                    
2050ea0c-0f62-434c-adc4-af3e0b997d94 accepted for retrieval                                                                                                                                                                    
a3313bf8-5070-40f7-aab3-16dc1e347d0d accepted for retrieval                                                                                                                                                                    
d54dee6f-d1d6-40ea-909b-2c5af9354d5a accepted for retrieval                                                                                                                                                                    
99e93043-a721-4560-ba97-6b23c4e1907a accepted for retrieval                                                                                                                                                                    
692d45de-4d1e-41df-b46c-2b41d8dc1d65 accepted for retrieval                                                                                                                                                                    
Download will resume from existing incomplete file S2A_MSIL2A_20190312T200221_N0213_R128_T14XML_20191022T090931.zip.incomplete.                                                                                                
2050ea0c-0f62-434c-adc4-af3e0b997d94 retrieval from LTA completed                                                                                                                                                              
96250b2d-d46e-4444-bfb7-516bd9068f35 retrieval from LTA completed                                                                                                                                                              
Downloading S2A_MSIL2A_20190312T200221_N0213_R128_T14XML_20191022T090931: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████| 602M/602M [05:12<00:00, 1.41MB/s]
Download will resume from existing incomplete file S2A_MSIL2A_20190312T200221_N0213_R128_T13XEF_20191022T090931.zip.incomplete.                                                                                                
LTA retrieval:  22%|█████████████████████████████████████▎                                                                                                                                  | 2/9 [05:17<11:35, 99.35s/product]
Downloading products:  35%|████████████████████████████████████████████████████████                                                                                                        | 7/20 [05:17<21:30, 99.23s/product]
Downloading S2A_MSIL2A_20190312T200221_N0213_R128_T13XDG_20191022T090931:  61%|███████████████████████████████████████████████████████████████████                                           | 425M/697M [08:43<07:18, 620kB/s]
Downloading S2A_MSIL2A_20190312T200221_N0213_R128_T13XDC_20191022T090931:  66%|████████████████████████████████████████████████████████████████████████▎                                     | 482M/734M [08:43<05:12, 805kB/s]
Downloading S2A_MSIL2A_20190312T200221_N0213_R128_T14XNL_20191022T090931:  89%|█████████████████████████████████████████████████████████████████████████████████████████████████▋            | 549M/619M [08:42<01:17, 904kB/s]
Downloading S2A_MSIL2A_20190312T200221_N0213_R128_T13XEF_20191022T090931:  27%|█████████████████████████████▍                                                                                | 180M/675M [03:27<09:30, 868kB/s]
... snip ...
Successfully downloaded 20/20 products.
```

Comments and reviews are welcome. I hope you forgive me the massive size of this PR.
To make it a bit easier to keep track of things I created a temporary branch that does not include the moving of functionality under `Downloader` [here](https://github.com/sentinelsat/sentinelsat/compare/main...valgur:feature/download_all-rewrite-wo-reorg?expand=1).

j08lue reviewed (approved):
This is a great code improvement, splitting out the download logic into its own class, explicitly declaring and handling status codes for products, and merging in the nodes filtering part. Great you managed to do all that backwards-compatible!

Great idea to use a semaphore for the upstream API requests. I am unable to see through, whether there is risk for deadlock or some other race conditions, but I guess only real world usage can really show. The old implementation does not work anymore with the radical rate limit of SciHub, so this is our very best bet (and a fine one) to make our client work again. I would say merge this and let the users tell us whether their issues are resolved or they experience new ones.

Up to you whether you want to follow my pointers regarding style/interface before merging.

valgur commented on sentinelsat/sentinel.py:
```diff
@@ -502,99 +546,30 @@ def download(self, id, directory_path=".", checksum=True, **kwargs):
-                raise InvalidChecksumError("File corrupt: checksums do not match")
-        # Download successful, rename the temporary file to its proper name
-        shutil.move(temp_path, path)
+        downloader = Downloader(self, node_filter=nodefilter, verify_checksum=checksum)
```
That's a good suggestion and I considered doing it that way, but it would cause issues if the download-related functions were called in parallel, since the configuration params are shared. A reasonable compromise might be to make it a member of `SentinelAPI` and `copy()` the downloader instance each time it is used, so the config can be safely modified.

j08lue replied:
Or that you pass in the class during `__init__` but it gets instantiated in `SentinelAPI.download`?

valgur replied:
I went with the `copy()` approach.

valgur commented:
TODO:
- [ ] `--fail-fast` has not been added to docs.
- [ ] `SentinelProductAPI` should be replaced with just `SentinelAPI` in the Sphinx docs.

valgur commented:
Thanks for the review and thanks for putting up with such a mammoth PR, @j08lue!

I made the suggested changes and added two new notable ones:
* The rate of download retries is now limited by `dl_retry_delay`, defaulting to 10 seconds. It really helps when the server has a transient error or a rate limit is being hit.
* *All* requests are now limited with the semaphore. Don't want to take any chances with SciHub :grin:. Deadlocks should be impossible since the scope of the semaphores is strictly limited to the `requests.get()`.

Also:
* Converted the new `download_all()` params into `SentinelAPI` attributes.
* Tweaked the download logging when using a node filter.

j08lue reviewed (approved):
Great work, which greatly improves the grown codebase and will hopefully get us back on track of having a working client for the Copernicus Open Access Hub with all its cumbersome LTA specialties.

joooeey commented:
What's the rationale for increasing `n_concurrent_dl` from 2 to 4? Did ESA change the restriction on their side? I can't find any mention of that anywhere.

joooeey commented:
I found the answer here: https://github.com/sentinelsat/sentinelsat/issues/488#issuecomment-851511217
ESA changed the limit earlier this year without announcing it.
